src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs

```diff
@@ -644,6 +644,88 @@ protected override void ApplyAnalyzerReferenceRemoved(ProjectId projectId, Analy
             }
         }
 
+        protected override void ApplyProjectChanges(ProjectChanges projectChanges)
+        {
+            base.ApplyProjectChanges(projectChanges);
+
+            var changedMappedDocuments = projectChanges.GetChangedDocuments()
+                .Where(id => ShouldApplyChangesToMappedDocuments(id))
+                .ToImmutableArray();
+            if (changedMappedDocuments.Any())
+            {
+                ApplyChangesForMappedDocuments(projectChanges, changedMappedDocuments);
+            }
+
+            bool ShouldApplyChangesToMappedDocuments(DocumentId id)
+            {
+                var document = this.CurrentSolution.GetDocument(id);
+                // Only consider files that are mapped and that we are unable to apply changes to.
+                return document?.Services?.GetService<ISpanMappingService>() != null && document?.CanApplyChange() == false;
+            }
+        }
+
+        private void ApplyChangesForMappedDocuments(ProjectChanges projectChanges, ImmutableArray<DocumentId> changedMappedDocuments)
+        {
+            // Get the original text changes from all documents and call the span mapping service to get span mappings for the text changes.
+            var mappedSpansPerDocument = GetTextChangesAndMappedSpans(projectChanges, changedMappedDocuments);
+
+            // Create mapped text changes using the mapped spans and original text changes.
+            using var _ = ArrayBuilder<(string FileName, TextChange TextChange)>.GetInstance(out var mappedChangesBuilder);
+            foreach (var mappedSpans in mappedSpansPerDocument)
+            {
+                var newDoc = projectChanges.NewProject.GetDocument(mappedSpans.Key)!;
+                MapTextChanges(newDoc, mappedSpans.Value.MappedSpans, mappedSpans.Value.OriginalTextChanges, mappedChangesBuilder);
+            }
```

Comment:
I'm not following this structure here: it seems GetTextChangesAndMappedSpans and MapTextChanges could both be inlined, and then the code unified and it'd end up simpler. Otherwise it seems like we're having to contort to some icky structures like the `Dictionary<DocumentId, (ImmutableArray<MappedSpanResult> MappedSpans, ImmutableArray<TextChange> OriginalTextChanges)>` for no real reason that I can tell...?

Reply:
yup - was just doing something silly!